app/controllers/api/widget_sets_controller.rb

```diff
@@ -0,0 +1,134 @@
+module Api
+  class WidgetSetsController < BaseController
+    REQUIRED_FIELDS_TO_COPY = %w[name].freeze
+    ALLOWED_FIELDS = REQUIRED_FIELDS_TO_COPY + %w[group description guid read_only set_data].freeze
+    SET_DATA_COLS = %i[col1 col2 col3].freeze
+
+    def init_set_data(set_data)
+      set_data ||= {}
+      new_set_data ||= {}
+      new_set_data[:col1] = set_data['col1'] || []
+      new_set_data[:col2] = set_data['col2'] || []
+      new_set_data[:col3] = set_data['col3'] || []
+      new_set_data[:reset_upon_login] = !!set_data['reset_upon_login']
+      new_set_data[:locked] = !!set_data['locked']
+      new_set_data
+    end
+
+    def init_widget_set(data)
+      data['set_type']   = "MiqWidgetSet"
+      data['owner_type'] = "MiqGroup"
+      data['set_data']   = init_set_data(data['set_data'])
+      data
+    end
+
+    def validate_description(group_id, description)
+      widget_set_exists_in_group = MiqWidgetSet.exists?(:owner_id => group_id, :description => description)
+      raise ArgumentError, "Description(Tab Title) must be unique for this group" if widget_set_exists_in_group
+    end
+
+    def validate_widgets_in_set_data(set_data)
+      widget_ids = SET_DATA_COLS.map { |x| set_data[x] }.flatten.compact
+      raise ArgumentError, "One widget must be selected(set_data)" if widget_ids.empty?
+
+      filtered_widget_ids = MiqWidget.where(:id => widget_ids).pluck(:id)
+      unless filtered_widget_ids.count == widget_ids.count
+        raise ArgumentError, "Unable to find widget set ids: #{widget_ids - filtered_widget_ids} "
+      end
+    end
+
+    def create_resource(type, id = nil, data = nil)
+      raise_if_unsupported_fields_passed(data)
+
+      raise ArgumentError, "Name cannot contain \"|\"" if data['name'].index('|')
```

Comment:
Also should be a model validation.

Reply:
Agreed, API does not raise any ArgumentErrors today, those come from the model.